vowpalwabbit/array_parameters.h

```diff
@@ -107,6 +108,33 @@ class sparse_parameters
 
   bool not_null() { return (_weight_mask > 0 && !_map.empty()); }
 
+  std::unordered_map<uint64_t, std::bitset<32>> feature_bit_vector_sparse; // define the 32-bitset for each feature
+  struct tag_hash_info_sparse // define struct to store the information of the tag hash
+  {
+    uint64_t tag_hash;
+    bool is_set=false;
+  };
+
+  tag_hash_info_sparse tag_info_sparse;
```

Comment:
I think since this is defined inside the `sparse_parameters` class it can just be named `tag_hash_info` (same for `feature_bit_vector`. Also both variables should probably be defined with the rest of the private vars for this class and start with an underscore (i.e. `_tag_hash_info`) for consistency